% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/funclib_PolyAtailor.R
\name{geneAnno}
\alias{geneAnno}
\title{Tail quantitative by alignment}
\usage{
geneAnno(tailDF, refPath, bamdf, GFF, longRead)
}
\arguments{
\item{tailDF}{The tailScan dataframe.This parameter is required if the
sequence type is shortreads, and omitted if the sequence type is longreads.}

\item{refPath}{The path of Reference conversion table.}

\item{bamdf}{The output dataframe of the function tailMap. Include at least
read_num,chr,strand,coord.}

\item{GFF}{GFF file, can be GFF/GTF/GFF3, or TXDB comment package.It is
recommended to use the TXDB annotation package, but be careful about the
correspondence of the reference genome version.}

\item{longRead}{Boolean value.If your sequence type is longreads this
parameter is T, otherwise it is F, and the default is T.}
}
\value{
Added tail table of gene annotation information.Include at least
  read_num,tail,PAL,chr,strand,gene,gene_type,tailType,read_type and sample.
}
\description{
\code{geneAnno}Add genetic information to the tail after
  alignment.
}
\details{
This function annotates the gene information, including the gene ID
  and gene type, using the tail of the matched gene from the GFF file.
}
\examples{
BiocManager::install("TxDb.Mmusculus.UCSC.mm10.knownGene")
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
load("./data/output/GV1tailDF.RData")
load("./data/output/GV1tailMapre.RData")
AnnotedTails = geneAnno(tailDF=GV1tailDF,bamdf=GV1tailMapre,GFF=TxDb.Mmusculus.UCSC.mm10.knownGene,longRead=F)
}
